chore(core): update `raise` typings to match `send`

Enteleform opened the pull request:
Fixes #2710

`raise` was programmatically handling `send`-shaped events, including callbacks; however, it wasn't typed properly to signal that it was capable of doing so.

I merged the relevant parts of the `send` function signature with `raise`, and refactored the test into a more concise implementation as per [**this discussion**](https://discord.com/channels/795785288994652170/801127053092716596/894393154743197717) with @davidkpiano on Discord.

davidkpiano commented on packages/core/src/actions.ts:
```diff
@@ -152,11 +152,16 @@ export function toActivityDefinition<TContext, TEvent extends EventObject>(
+export function raise<
+  TContext,
+  TEvent extends EventObject,
+  TSentEvent extends EventObject = AnyEventObject
```
This shouldn't be necessary; it should be the same as `TEvent` for raising events

Enteleform replied:
`TSentEvent` ?

so like; `SendExpr<TContext, TEvent, TEvent>`, etc. instead ?

Enteleform replied:
Haven't dug that deep into the internal differences between `send` and `raise`, just assumed that since non-string events are being forwarded to `send` that using the same typings would provide the same IntelliSense & whatnot.

davidkpiano replied:
> `TSentEvent` ?
> 
> so like; `SendExpr<TContext, TEvent, TEvent>`, etc. instead ?

Yes, like this, I'm pretty sure

Enteleform commented:
@Andarist Thanks for the thorough review & details! Sorry for the delay, been swamped lately. I just took another shot at this after checking out your TS playground example, but haven't been able to figure out a solution that addresses the issues that you highlighted just yet.

Might have to come back to this after I wrap up my current project, been kind of short on time & the mental energy required for type programming (*been spending way too much time in React render optimization land lately 💀*).

Andarist commented:
Superseded by https://github.com/statelyai/xstate/pull/3393
